extensions/data-workspace/src/services/workspaceService.ts

```diff
@@ -13,11 +14,45 @@ import * as constants from '../common/constants';
 
 const WorkspaceConfigurationName = 'dataworkspace';
 const ProjectsConfigurationName = 'projects';
+const TempProject = 'tempProject';
 
 export class WorkspaceService implements IWorkspaceService {
 	private _onDidWorkspaceProjectsChange: vscode.EventEmitter<void> = new vscode.EventEmitter<void>();
 	readonly onDidWorkspaceProjectsChange: vscode.Event<void> = this._onDidWorkspaceProjectsChange?.event;
 
+	constructor(private _context: vscode.ExtensionContext) {
+	}
+
+	/**
+	 * Load any temp project that needed to be loaded before the extension host was restarted
+	 * which would happen if a workspace was created in order open or create a project
+	 */
+	async loadTempProject(): Promise<void> {
+		const tempProjectFile = this._context.globalState.get(TempProject);
+
+		if (tempProjectFile && vscode.workspace.workspaceFile) {
+			// add project to workspace now that the workspace has been created and saved
+			await this.addProjectsToWorkspace([vscode.Uri.file(<string>tempProjectFile)]);
+			await this._context.globalState.update(TempProject, undefined);
+		}
+	}
+
+	/**
+	 * Creates a new workspace in the same folder as the project. Because the extension host gets restared when
+	 * a new workspace is created and opened, the project needs to be saved as the temp project that will be loaded
+	 * when the extension gets restarted
+	 * @param projectFileFsPath project to add to the workspace
+	 */
+	async CreateNewWorkspaceForProject(projectFileFsPath: string): Promise<void> {
+		// save temp project
+		await this._context.globalState.update(TempProject, projectFileFsPath);
+
+		// create a new workspace - the workspace file will be created in the same folder as the project
+		const workspaceFileFsPath = vscode.Uri.file(path.join(path.dirname(projectFileFsPath), `${path.parse(projectFileFsPath).name}.code-workspace`)).fsPath;
+		const projectFolder = vscode.Uri.file(path.dirname(projectFileFsPath));
+		await azdata.workspace.createWorkspace(projectFolder, workspaceFileFsPath);
```

Comment:
please also add a unit test for the scenario when creating workspace is required